src/esp/gfx/PbrDrawable.cpp

```diff
@@ -0,0 +1,211 @@
+    shader_->bindBaseColorTexture(*materialData_->baseColorTexture);
+  }
+
+  if ((flags_ & PbrShader::Flag::RoughnessTexture) &&
+      materialData_->roughnessTexture) {
+    shader_->bindRoughnessTexture(*materialData_->roughnessTexture);
+  }
+
+  if ((flags_ & PbrShader::Flag::MetallicTexture) &&
+      materialData_->metallicTexture) {
+    shader_->bindMetallicTexture(*materialData_->metallicTexture);
+  }
+
+  if ((flags_ & PbrShader::Flag::NoneRoughnessMetallicTexture) &&
+      materialData_->noneRoughnessMetallicTexture) {
+    shader_->bindNoneRoughnessMetallicTexture(
+        *materialData_->noneRoughnessMetallicTexture);
+  }
+
+  if ((flags_ & PbrShader::Flag::OcclusionRoughnessMetallicTexture) &&
+      materialData_->occlusionRoughnessMetallicTexture) {
+    shader_->bindOcclusionRoughnessMetallicTexture(
+        *materialData_->occlusionRoughnessMetallicTexture);
+  }
+
+  if ((flags_ & PbrShader::Flag::NormalTexture) &&
+      materialData_->normalTexture) {
+    shader_->bindNormalTexture(*materialData_->normalTexture);
+  }
+
+  if ((flags_ & PbrShader::Flag::EmissiveTexture) &&
+      materialData_->emissiveTexture) {
+    shader_->bindEmissiveTexture(*materialData_->emissiveTexture);
+  }
+
+  if (materialData_->textureMatrix != Mn::Matrix3{}) {
+    shader_->setTextureMatrix(materialData_->textureMatrix);
+  }
+
+  shader_->draw(mesh_);
+}
+
+Mn::ResourceKey PbrDrawable::getShaderKey(Mn::UnsignedInt lightCount,
+                                          PbrShader::Flags flags) const {
+  return Corrade::Utility::formatString(
+      SHADER_KEY_TEMPLATE, lightCount,
+      static_cast<PbrShader::Flags::UnderlyingType>(
+          PbrShader::generateCorrectFlags(flags)));
+}
+
+PbrDrawable& PbrDrawable::updateShader() {
+  unsigned int lightCount = lightSetup_->size();
+  if (!shader_ || shader_->lightCount() != lightCount ||
+      shader_->flags() != flags_) {
+    // if the number of lights or flags have changed, we need to fetch a
+    // compatible shader
+    shader_ = shaderManager_.get<Mn::GL::AbstractShaderProgram, PbrShader>(
+        getShaderKey(lightCount, flags_));
+
+    // if no shader with desired number of lights and flags exists, create one
+    if (!shader_) {
+      shaderManager_.set<Mn::GL::AbstractShaderProgram>(
+          shader_.key(), new PbrShader{flags_, lightCount},
+          Mn::ResourceDataState::Final, Mn::ResourcePolicy::ReferenceCounted);
+    }
+
+    CORRADE_INTERNAL_ASSERT(shader_ && shader_->lightCount() == lightCount &&
+                            shader_->flags() == flags_);
+  }
+
+  return *this;
+}
+
+// update every light's color, intensity, range etc.
+PbrDrawable& PbrDrawable::updateShaderLightParameters() {
+  constexpr float dummyRange = Mn::Constants::inf();
+  std::vector<float> ranges;
+  std::vector<Mn::Color3> colors;
+  colors.reserve(lightSetup_->size());
+  for (unsigned int iLight = 0; iLight < lightSetup_->size(); ++iLight) {
+    ranges.push_back(dummyRange);
```

Comment:
Remember my [brainfart comment](https://github.com/facebookresearch/habitat-sim/pull/812#discussion_r497328490)?  Just to finish that -- in Magnum's builtin shaders, the shader has reasonable defaults right after its construction. So if you do

```cpp
Shaders::Whichever shader;
// and maybe set the transformation matrices
shader.draw(mesh);
```

it will always do *something* and won't result in a black screen -- if you don't set any light parameters, it'll use reasonable defaults, it you don't set any color, it'll use white, if you don't set any transformation, it'll use identities etc.

Now, it's up to you if you want to have the same "reasonable defaults" behavior in Habitat or not (at which point, you wouldn't need to set implicit infinity ranges because those are there by default). I could imagine that it could help to whoever developer comes next into the rendering code here, or if it gets to the point where Python users are instantiating their own meshes and using shaders, this might be a lot of help, but maybe also not. Up to you :)

Reply:
Now, this time, I prefer to encourage it to fail earlier. 

> won't result in a black screen 

I would give them the black screen so they immediately know they forget to setup the lights for example.

>  it'll use reasonable defaults

Again, if the users forget to setup the lights, and it is not black (due to your default reasonable defaults), they may think the shader code has problem because the rendering results do not match what they expect (for example, they expect the object is red, but now it is white).

Reply:
I see we have vastly different opinions on this topic, but I'll try one last time -- and if our opinions still diverge after this point, I'd say we stop and "agree to disagree" to save ourselves time :) 

I used to not have any defaults in builtin shaders, but since I have them, user experience improved a lot. Based on my experience with answering 100s of questions in Magnum support channels over the years:

- Having a black / empty render can have way too many reasons (wrong triangle winding, renderer state, depth buffer not cleared, model missing normals, model having flipped normals, camera clipping planes, wrong transformation matrices...) and because the problem space is so big, I'm taking every chance to make the space smaller, because debugging "my object is white but my lights are red" is much easier than "i tried everything but all i see is _still_ black".
- The more complex shaders have a ton of parameters and users jumping into it for the first time will just try to set the parameters they know (like transformation/projection matrices) and see what comes out. If they get something, it'll be a reward nudging them to continue exploring further with colors, lights, textures etc. If they get a black screen, the fun ends there.
- If a shader gets a new parameter (such as the normal map scale that recently arrived in Phong) and it doesn't have a reasonable default, existing code will be broken (normal mapping doing nothing) until someone notices (which might also be never). This is probably the strongest point for reasonable defaults as it applies even when your team is full of expert veterans who never suffered from a black screen render in their lifetime.